File: 0383-ransom-note/0383-ransom-note.py
class Solution(object):
    def canConstruct(self, ransomNote, magazine):
        """
        :type ransomNote: str
        :type magazine: str
        :rtype: bool
        """
        hash=[0]*128
        for i in magazine:
            hash[ord(i)]+=1
        for i in ransomNote:
            hash[ord(i)]-=1
        for i in hash:
            if i<0:
                return False
        return True

        

File: 0383-ransom-note/0383-ransom-note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        int[] hash = new int[128];// space complexity is O(n) because even though we initialized new array the size is 128 we declared so no matter how long the input is it will always come under this 128 suppose if put size as n (dynamic array) then it can be O(n) but here since we declared the size 128 it is constant
        for(char ch:magazine.toCharArray()){
            hash[ch-'0']++;
        }
        for(char ch:ransomNote.toCharArray()){
            hash[ch-'0']--;
        }
        for(int i:hash){
            if(i<0){
                return false;
            }
        }
        return true;
        

    }
}